src/meta/src/stream/scale.rs

```diff
@@ -0,0 +1,827 @@
+                    }
+                }
+
+                let worker_id = parallel_unit_id_to_worker_id
+                    .get(new_parallel_unit_id)
+                    .unwrap();
+
+                let worker = worker_nodes.get(worker_id).unwrap();
+
+                node_actors
+                    .entry(*worker_id)
+                    .or_default()
+                    .push(new_actor.clone());
+
+                broadcast_node_ids.insert(*worker_id);
+
+                actor_infos_to_broadcast.insert(
+                    *new_actor_id,
+                    ActorInfo {
+                        actor_id: *new_actor_id,
+                        host: worker.host.clone(),
+                    },
+                );
+            }
+        }
+
+        for node_id in &broadcast_node_ids {
+            let node = worker_nodes.get(node_id).unwrap();
+            let client = self.client_pool.get(node).await?;
+
+            client
+                .to_owned()
+                .broadcast_actor_info_table(BroadcastActorInfoTableRequest {
+                    info: actor_infos_to_broadcast.values().cloned().collect(),
+                })
+                .await?;
+        }
+
+        for (node_id, stream_actors) in &node_actors {
+            let node = worker_nodes.get(node_id).unwrap();
+            let client = self.client_pool.get(node).await?;
+            let request_id = Uuid::new_v4().to_string();
+            let request = UpdateActorsRequest {
+                request_id,
+                actors: stream_actors.clone(),
+                hanging_channels: node_hanging_channels.remove(node_id).unwrap_or_default(),
+            };
+
+            client.to_owned().update_actors(request).await?;
+        }
+
+        for (node_id, hanging_channels) in node_hanging_channels {
+            let node = worker_nodes.get(&node_id).unwrap();
+
+            let client = self.client_pool.get(node).await?;
+            let request_id = Uuid::new_v4().to_string();
+
+            client
+                .to_owned()
+                .update_actors(UpdateActorsRequest {
+                    request_id,
+                    actors: vec![],
+                    hanging_channels,
+                })
+                .await?;
+        }
+
+        for (node_id, stream_actors) in node_actors {
+            let node = worker_nodes.get(&node_id).unwrap();
+            let client = self.client_pool.get(node).await?;
+            let request_id = Uuid::new_v4().to_string();
+
+            client
+                .to_owned()
+                .build_actors(BuildActorsRequest {
+                    request_id,
+                    actor_id: stream_actors
+                        .iter()
+                        .map(|stream_actor| stream_actor.actor_id)
+                        .collect(),
+                })
+                .await?;
+        }
+
+        // then, we collect all available upstream
+        let mut upstream_dispatchers = HashMap::new();
```

Comment:
```suggestion
        let mut actor_upstream_dispatchers_map = HashMap::new();
```

Seems this logic can also be extracted?